### PR TITLE
 BUG 1869497:   fix invalid ownerrefernce for localvolumediscoveryresult CR

### PR DESCRIPTION
**Pull request by sp98**
LocalVolumeDiscovery CR uses ownerrefernce>kind as LocalVolumeDisovery but uses the ownerrefernce UUID of daemonset POD

This PR changes the owner references of the LocalVolumeDiscoveryResult objects to LocalVolumeDiscovery object. The reconcile method deletes the orphan results on every run. 

Signed-off-by: Santosh Pillai <sapillai@redhat.com>

### CODE REVIEW

File: pkg/diskmaker/discovery/sync_result.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ func newDiscoveryResultInstance(nodeName, namespace, parentObjName, parentObjUID
 			OwnerReferences: []metav1.OwnerReference{
 				{
 					APIVersion: v1alpha1.SchemeGroupVersion.String(),
-					Kind:       "LocalVolumeDiscovery",
```
Comment (gnufied) [inline pkg/diskmaker/discovery/sync_result.go:33]: I do not understand why should discovery results object have owneref to pod or daemonset. Shouldn't this be `LocalVolumeDiscovery` object and its UUID.

Comment (sp98) [reply pkg/diskmaker/discovery/sync_result.go:33]: I might be wrong, but how about the following scenario:
1-  User startts Discovery on Node1 and Node2.
2- User updates the Discovery CR to run it only on Node1. 

After Step 2,  daemonset pod running on `Node2` is gone so corresponding discovery result CR will be deleted. 
But if the Discovery Result CR was dependent on `LocalVolumeDiscovery`, then after step 2, daemonset pod will be deleted but the discovery result CR (on Node1) won't be deleted. 

So in short, if we have LocalVolumeDiscovery and its UUID as owner, the results will be deleted only when user deletes the LocalVolumeDiscovery CR complete and not when user updates it.

Comment (gnufied) [reply pkg/diskmaker/discovery/sync_result.go:33]: I don't think it is a good idea to depend on ownereref mechanism for cleaning up leftover result objects. We already have established pattern of verifying if node selector includes certain node and may be we should periodically reconcile and delete leftover discovery result objects and use LocalVolumeDiscovery object as owneref.

Comment (sp98) [reply pkg/diskmaker/discovery/sync_result.go:33]: Identifying `leftover discovery result objects` approach might get complicated if we decide to go with multiple `LocalVolumeDiscovery` objects in the future.  Right now, we mandate the user to create only a single `LocalVolumeDiscovery` object. 

With multiple `LocalVolumeDisovery` objects, a node might be removed from one object's nodeSelector, but it still might be part of another object's nodeSelector.

Comment (sp98) [reply pkg/diskmaker/discovery/sync_result.go]: @gnufied Updated the owner reference.  LocalVolumeDiscoveryResults have owner reference of LocalVolumeDiscovery. Also deleting any orphan LocalVolumeDiscoveryResult on every reconcile.

File: pkg/controller/localvolumediscovery/localvolumediscovery_controller.go [hunk 2]
```diff
@@ -283,13 +328,15 @@ func getOwnerRefs(cr *localv1alpha1.LocalVolumeDiscovery) []metav1.OwnerReferenc
...
-	return corev1.EnvVar{
-		Name: "UID",
-		ValueFrom: &corev1.EnvVarSource{
-			FieldRef: &corev1.ObjectFieldSelector{
-				FieldPath: "metadata.uid",
-			},
+func getEnvVars(objName, uid string) []corev1.EnvVar {
+	return []corev1.EnvVar{
+		{
+			Name:  "UID",
```
Comment (gnufied) [inline pkg/controller/localvolumediscovery/localvolumediscovery_controller.go:334]: Can this environment variable be suitable named? `UID` as a generic name can conflict with other things.

File: pkg/diskmaker/discovery/sync_result.go [hunk 3]
```diff
@@ -49,23 +49,14 @@ func (discovery *DeviceDiscovery) ensureDiscoveryResultCR() error {
 	nodeName := os.Getenv("MY_NODE_NAME")
 	namespace := os.Getenv("WATCH_NAMESPACE")
 	parentObjUID := os.Getenv("UID")
-	parentObjName := os.Getenv("POD_NAME")
+	parentObjName := os.Getenv("DISCOVERY_OBJECT_NAME")
 	if nodeName == "" || namespace == "" || parentObjUID == "" || parentObjName == "" {
```
Comment (gnufied) [inline pkg/diskmaker/discovery/sync_result.go:53]: should we have a unit test for proper owner ref on discovery result objects?

File: pkg/controller/localvolumediscovery/localvolumediscovery_controller.go [hunk 4]
```diff
@@ -202,6 +217,36 @@ func (r *ReconcileLocalVolumeDiscovery) updateDiscoveryStatus(instance *localv1a
...
+
+	for _, discoveryResult := range discoveryResultList.Items {
+		nodeName := discoveryResult.Spec.NodeName
+		node := &corev1.Node{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node)
+		if err != nil {
+			return fmt.Errorf("failed to get instance of node %q", nodeName)
+		}
+
+		matches := corev1helper.MatchNodeSelectorTerms(instance.Spec.NodeSelector.NodeSelectorTerms,
```
Comment (gnufied) [inline pkg/controller/localvolumediscovery/localvolumediscovery_controller.go:235]: this causes LSO to crash, if no nodeSelector was specified in CR..

Comment (gnufied) [reply pkg/controller/localvolumediscovery/localvolumediscovery_controller.go:235]: please ensure that we have a test case for this case.

Comment (sp98) [reply pkg/controller/localvolumediscovery/localvolumediscovery_controller.go:246]: Thanks. Added a check to skip when no NodeSelector is specified [here](https://github.com/openshift/local-storage-operator/blob/9c72cb708080ed6693dc090f52ed3282ca3fb6cc/pkg/controller/localvolumediscovery/localvolumediscovery_controller.go#L227).  Added unit test for this case here [https://github.com/openshift/local-storage-operator/blob/9c72cb708080ed6693dc090f52ed3282ca3fb6cc/pkg/controller/localvolumediscovery/localvolumediscovery_controller_test.go#L275](url)

File: pkg/diskmaker/discovery/sync_result.go (hunk 3 above)
Comment (sp98) [reply pkg/diskmaker/discovery/sync_result.go:53]: We do have some existing tests for that [here](https://github.com/openshift/local-storage-operator/blob/9c72cb708080ed6693dc090f52ed3282ca3fb6cc/pkg/diskmaker/discovery/sync_result_test.go#L144-L145) Will these suffice?

File: pkg/controller/localvolumediscovery/localvolumediscovery_controller.go (hunk 2 above)
Comment (sp98) [reply pkg/controller/localvolumediscovery/localvolumediscovery_controller.go]: Updated name to `DISCOVERY_OBJECT_UID`
